src/index.js

```diff
@@ -9,10 +9,294 @@ OF ANY KIND, either express or implied. See the License for the specific languag
 governing permissions and limitations under the License.
 */
 
-const AEMHeadless = require('../src/main')
-const getToken = require('./utils/auth')
+const { SDKError, SDKErrorWrapper } = require('./utils/errors')
+const { AEM_GRAPHQL_ACTIONS } = require('./utils/config')
 
-module.exports = {
-  AEMHeadless,
-  getToken
+/**
+ * This class provides methods to call AEM GraphQL APIs.
+ * Before calling any method initialize the instance
+ * with GraphQL endpoint, GraphQL serviceURL and auth if needed
+ */
+class AEMHeadless {
+  /**
+   * Constructor.
+   *
+   * If param is a string, it's treated as AEM server URL, default GraphQL endpoint is used.
+   * For granular params, use config object
+   *
+   * @param {string|object} config - Configuration object, or AEM server URL string
+   * @param {string} [config.serviceURL] - AEM server URL
+   * @param {string} [config.endpoint] - GraphQL endpoint
+   * @param {(string|Array)} [config.auth] - Bearer token string or [user,pass] pair array
+   * @param {object} [config.fetch] - Fetch instance - required for NodeJS only, eg node-fetch/cross-fetch
+   */
+  constructor (config) {
+    let endpoint = AEM_GRAPHQL_ACTIONS.endpoint
+    let serviceURL = AEM_GRAPHQL_ACTIONS.serviceURL
+
+    if (typeof config === 'string') {
+      serviceURL = config
+    } else {
+      serviceURL = config.serviceURL || serviceURL
+      endpoint = config.endpoint || endpoint
+      this.auth = config.auth
+    }
+
+    this.serviceURL = this.__getDomain(serviceURL)
+    this.endpoint = this.__getPath(endpoint)
+    this.fetch = this.__getFetch(config.fetch)
+  }
+
+  /**
+   * Returns a Promise that resolves with a POST request JSON data.
+   *
+   * @param {string} query - the query string
+   * @param {object} [options={}] - additional POST request options
+   * @returns {Promise<any>} - the response body wrapped inside a Promise
+   */
+  runQuery (query, options = {}) {
+    return this.__handleRequest(this.endpoint, JSON.stringify({ query }), options)
+  }
+
+  /**
+   * Returns a Promise that resolves with a PUT request JSON data.
+   *
+   * @param {string} query - the query string
+   * @param {string} path - AEM path to save query, format: configuration_name/endpoint_name
+   * @param {object} [options={}] - additional PUT request options
+   * @returns {Promise<any>} - the response body wrapped inside a Promise
+   */
+  persistQuery (query, path, options = {}) {
+    const url = `${AEM_GRAPHQL_ACTIONS.persist}/${path}`
+    return this.__handleRequest(url, query, { method: 'PUT', ...options })
+  }
+
+  /**
+   * Returns a Promise that resolves with a GET request JSON data.
+   *
+   * @param {object} [options={}] - additional GET request options
+   * @returns {Promise<any>} - the response body wrapped inside a Promise
+   */
+  listPersistedQueries (options = {}) {
```

Comment:
should have `async` keyword since it returns a Promise (handleRequest)

Reply:
Updated